optuna/study/study.py

```diff
@@ -1018,23 +992,169 @@ def _tell(
+
+        if state is None and values is None:
+            state = TrialState.FAIL
+            message_values_and_state_not_specified = (
+                "Study.tell is called with `values` and `state` set to None but "
+                "you have to specify either `values` or `state` at least. "
+                "`values` is required if a trial finishes successfully or "
+                "`state` is required if a trial is fails or pruned."
+            )
+            if not suppress_warning:
+                warnings.warn(message_values_and_state_not_specified)
+            else:
+                warning_message = message_values_and_state_not_specified
+
+        if state == TrialState.PRUNED:
+            # Register the last intermediate value if present as the value of the trial.
+            # TODO(hvy): Whether a pruned trials should have an actual value can be discussed.
+            assert values is None
+
+            last_step = frozen_trial.last_step
+            if last_step is not None:
+                values = [frozen_trial.intermediate_values[last_step]]
+
+        if values is not None:
+            values, values_conversion_failure_message = _check_and_convert_to_values(
+                len(self.directions), values, trial_number
+            )
+
+            if state is None:
+                if values_conversion_failure_message is None:
+                    state = TrialState.COMPLETE
+
+                # This path is reached when the objective returns invalid data
+                # including NaN or non-numerical points during Study.optimize.
+                else:
+                    state = TrialState.FAIL
+                    if not suppress_warning:
+                        warnings.warn(values_conversion_failure_message)
+                    else:
+                        warning_message = values_conversion_failure_message
+
+            # When called from `Study.optimize` and `state` is pruned, the given `values` contains
+            # the intermediate value with the largest step so far. In this case, the value is
+            # allowed to be NaN and errors should not be raised.
+            elif state != TrialState.PRUNED and values_conversion_failure_message is not None:
+                raise ValueError(values_conversion_failure_message)
+
+        assert state is not None
+
+        try:
+            # Sampler defined trial post-processing.
+            study = pruners._filter_study(self, frozen_trial)
+            self.sampler.after_trial(study, frozen_trial, state, values)
+        except Exception:
+            raise
+        finally:
+            if values is not None:
+                self._storage.set_trial_values(trial_id, values)
+
+            self._storage.set_trial_state(trial_id, state)
+
+        frozen_trial = copy.deepcopy(self._storage.get_trial(trial_id))
+        if warning_message is not None:
+            frozen_trial.set_system_attr("study_tell_warning", warning_message)
```

Comment:
Ah, so it it's not persisted in the storage. Not what I first expected but sounds fair.